File: algorithms/java/230_kth_smallest_element_in_a_BST/kth_smallest_element_in_a_BST.java
import java.util.Scanner;

import tree.BinTree;
import tree.BinTree.TreeNode;

/**
 * @author Peng
 * Problem Description:
 *      Given a binary search tree, write a function kthSmallest 
 *      to find the kth smallest element in it.
 *      
 *      Note: 
 *          You may assume k is always valid, 1 ¡Ü k ¡Ü BST's total elements.
 *          
 *      Follow up:
 *          What if the BST is modified (insert/delete operations) often and you need to
 *          find the kth smallest frequently? How would you optimize the kthSmallest routine?
 *      
 *      Hint:
 *          Try to utilize the property of a BST.
 *          What if you could modify the BST node's structure?
 *          The optimal runtime complexity is O(height of BST).
 */
public class kth_smallest_element_in_a_BST {

    /**
     * Solution 1:
     *      we count the child node numbers and compare to choose sides
     *      and recursive
     * @param root
     * @param k
     * @return 
     */
    public static int kthSmallest_1(TreeNode root, int k) {
        int n = count(root.left);
        
        if(n >= k)  return kthSmallest_1(root.left, k);
        if(n + 1 == k)  return root.val;
        return kthSmallest_1(root.right, k - 1 - n);
    }
    /**
     * count the descendant node number
     * @param root
     * @return descendant node number
     */
    private static int count(TreeNode root) {
        if(root == null) return 0;
        return 1 + count(root.left) + count(root.right);
    }
    
    public static void main(String[] args) {
        // create a binary tree
        Integer[] data = {1, null, 5, null, null, 3, null};
        BinTree bt = new BinTree();
        bt.bTreeRoot = bt.createTree(data, 0, 7);
        
        Scanner scanner = new Scanner(System.in); 
        while (scanner.hasNext()) {
            int n = scanner.nextInt();
            System.out.println("the k-th smallest element is: " + kthSmallest_1(bt.bTreeRoot, n));
        }

        scanner.close();
    }

}

File: algorithms/python/230_kth_smallest_element_in_a_bst/kth_smallest_element_in_a_bst.py
# coding: utf8
# python 2.7
"""
Link: https://leetcode-cn.com/problems/kth-smallest-element-in-a-bst

Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

Note:
    You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

Example:
    Input: root = [5,3,6,2,4,null,null,1], k = 3
           5
          / \
         3   6
        / \
       2   4
      /
     1
Output: 3

Follow up:
What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?
"""

import sys
sys.path.append("../")
from self_made_libs.binary_tree import BTreeHelper, TreeNode

class Solution(object):

    def kthSmallest(self, root, k):
        """
            on the base of v1, we can stop early as soon as we found the k-th item
            here we use iterative traversal
            now the space complexity reduce to O(1)
        """
        res = root.val   
        s = []
        p = root
        while k:
            while p:
                s.append(p)
                p = p.left
            p = s.pop()
            res = p.val
            p = p.right
            k -= 1

        return res

    def kthSmallest_v1(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        :idea: using in-order traversal, and the k-th item is the target
        :complexity: O(n) in time and O(n) in space
        """
        def help(root, res):
            if not root:
                return            
            help(root.left, res)
            res.append(root.val)
            help(root.right, res)

        arr = []
        help(root, arr)
        return arr[k-1]

if __name__ == '__main__':
    BT = BTreeHelper()
    S = Solution()
    arr = [5,3,6,2,4,None,None,1]
    root = BT.build_a_tree(arr)
    # BT.trav_pre(root)
    res = S.kthSmallest(root, 3)
    print res
